vectorselector/couchdb.cr
# Drhuffman12/upgrade to crystal 1.1.1 (and 1.2.0-dev) (#1) and expand CouchDB version checks

-[x] Add dependency of `crystal-lang/json_mapping.cr`/ (`JSON.mapping` was deprecated and moved to shard.)
  * https://github.com/crystal-lang/crystal/pull/9527
  * https://github.com/crystal-lang/json_mapping.cr

-[x] Add more environment variables for database connection settings with defaults.
  * For now, just for tests:
    ```
      un = ENV.keys.includes?("TEST_DB_UN") ? ENV["TEST_DB_UN"] : "admin"
      pw = ENV.keys.includes?("TEST_DB_PW") ? ENV["TEST_DB_PW"] : "password"
      ip = ENV.keys.includes?("TEST_DB_IP") ? ENV["TEST_DB_IP"] : "localhost"
      port = ENV.keys.includes?("TEST_DB_PORT") ? ENV["TEST_DB_PORT"] : "5984"
      couchdb_url = ENV.keys.includes?("TEST_DB") ? ENV["TEST_DB"] : "http://#{un}:#{pw}@#{ip}:#{port}"
    ```

-[x] DEPRECATION: Use `ServerInfo#version` instead of `Vendor#version`.
  As of v3.x (or earlier?), a call to the server's root (e.g.: `GET http://127.0.0.1:5984/`)
    no longer returns `version` inside `vendor`.
  In v3.x, response will instead look like (as of v3.1.1):

  ```
  {
    "couchdb":"Welcome",
    "version":"3.X.X",
    "git_sha":"GIT_SHA_GOES_HERE",
    "uuid":"SOME_UUID_GOES_HERE",
    "features":["access-ready","partitioned","pluggable-storage-engines","reshard","scheduler"],
    "vendor":{"name":"The Apache Software Foundation"}
  }
  ```

-[x] Allow `3.x.x` versions of CouchDB.
  * Add `is_v2?` and `is_v3?` methods to `CouchDB::Response::ServerInfo`.
  * Tested locally w/ `3.x.x` and via CircleCI w/ `2.x.x`

-[x] Remove hard-coded `crystal` key/version from `shard.yml` (for now).
  * Tested (and passed)
    * locally w/ `Crystal 1.1.1 [6d9a1d583] (2021-07-26)`
    * via CircleCI w/ `Crystal 1.2.0-dev [cadc4b3c7] (2021-09-02)`

-[x] Bump version of this shard to `0.4.0`
  * Set `CouchDB::VERSION` based on shard version value in from `shard.yml` (to avoid out-of-sync issues).

---

`drhuffman12/ci_matrix CouchDB (2.x? and 3.x?)` (#2) adds CI test runs for matrix CouchDB (2.x? and 3.x?):

 - [x] Expand the CI to run the tests against a 'matrix' of version combo's made up of the:
    * latest 2 versions of CouchDB (2.x? and 3.x?)
